lib/zcaps.js

```diff
@@ -0,0 +1,201 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+const {CapabilityDelegation} = require('ocapld');
+const {EdvClient} = require('edv-client');
+const jsigs = require('jsonld-signatures');
+
+const {SECURITY_CONTEXT_V2_URL, sign, suites} = jsigs;
+const {Ed25519Signature2018} = suites;
+const SUPPORTED_KEY_TYPES = [
+  'Ed25519VerificationKey2018',
+  'Sha256HmacKey2019',
+  'X25519KeyAgreementKey2019'
+];
+
+// module API
+const api = {};
+module.exports = api;
+
+api.delegateCapability = async (
+  {signer, keystoreAgent, edvClient, request} = {}) => {
+  const {
+    invocationTarget, invoker, delegator, controller, referenceId,
+    allowedAction, caveat
+  } = request;
+  if(!(invocationTarget && typeof invocationTarget === 'object' &&
+    invocationTarget.type)) {
+    throw new TypeError(
+      '"invocationTarget" must be an object that includes a "type".');
+  }
+  // TODO: Look into requiring an invoker or controller on a zcap
+  let zcap = {
+    '@context': SECURITY_CONTEXT_V2_URL,
+    // use 128-bit random multibase encoded value
+    id: await api.id()
+  };
+  if(invoker) {
+    zcap.invoker = invoker;
+  }
+  if(controller) {
+    zcap.controller = controller;
+  }
+  if(delegator) {
+    zcap.delegator = delegator;
+  }
+  if(referenceId) {
+    zcap.referenceId = referenceId;
+  }
+  if(allowedAction) {
+    zcap.allowedAction = allowedAction;
+  }
+  if(caveat) {
+    zcap.caveat = caveat;
+  }
+  let {parentCapability} = request;
+  const {id: target, type: targetType, verificationMethod} = invocationTarget;
+  if(SUPPORTED_KEY_TYPES.includes(targetType)) {
+    if(!target) {
+      throw new TypeError(
+        '"invocationTarget.id" must be set for Web KMS capabilities.');
+    }
+    if(!verificationMethod) {
+      throw new TypeError('"invocationTarget.verificationMethod" is required.');
+    }
+    // TODO: fetch `target` from a key mapping document in the profile's
+    // edv to get public key ID to set as `referenceId`
+    zcap.invocationTarget = {
+      id: target,
+      type: targetType,
+      verificationMethod,
+    };
+    zcap.parentCapability = parentCapability || target;
+    zcap = await _delegate({zcap, signer});
+
+    return zcap;
+  } else if(targetType === 'urn:edv:document') {
+    zcap.invocationTarget = {
+      id: target,
+      type: targetType
+    };
+
+    if(target) {
+      // TODO: handle case where an existing target is requested
```

Comment:
Should this throw for now?  Just thinking about how it would be to debug this noop.

Reply:
We can throw for now.